prow/gerrit/reporter/reporter.go

```diff
@@ -62,33 +65,83 @@ func (c *Client) ShouldReport(pj *v1.ProwJob) bool {
 	}
 
 	// has gerrit metadata (scheduled by gerrit adapter)
-	return pj.ObjectMeta.Annotations["gerrit-id"] != "" &&
-		pj.ObjectMeta.Annotations["gerrit-instance"] != "" &&
-		pj.ObjectMeta.Labels["gerrit-revision"] != ""
+	if pj.ObjectMeta.Annotations[gerrit.GerritID] == "" ||
+		pj.ObjectMeta.Annotations[gerrit.GerritInstance] == "" ||
+		pj.ObjectMeta.Labels[gerrit.GerritRevision] == "" {
+		return false
+	}
+
+	// has other jobs on the same revision finished
+	selector := labels.Set{gerrit.GerritRevision: pj.ObjectMeta.Labels[gerrit.GerritRevision]}
+	pjs, err := c.lister.List(selector.AsSelector())
+	if err != nil {
+		logrus.WithError(err).Errorf("Cannot list prowjob with selector %v", selector)
+		return false
+	}
+
+	for _, pj := range pjs {
+		if pj.Status.State == v1.TriggeredState || pj.Status.State == v1.PendingState {
+			// not done yet
+			return false
+		}
+	}
+
+	return true
 }
 
 // Report will send the current prowjob status as a gerrit review
 func (c *Client) Report(pj *v1.ProwJob) error {
-	// TODO(krzyzacy): we should also do an aggregate report, as golang does in their repo
-	// see https://go-review.googlesource.com/c/go/+/132155
-	// also add ability to set code-review labels
-	// ref: https://github.com/kubernetes/test-infra/issues/9433
+	// If you are hitting here, which means the entire patchset has been finished :-)
 
-	if pj.Spec.Refs == nil {
-		return errors.New("no pj.Spec.Refs, not a presubmit job (should not happen?!)")
+	// list all prowjobs in the patchset
+	selector := labels.Set{gerrit.GerritRevision: pj.ObjectMeta.Labels[gerrit.GerritRevision]}
+	pjsOnRevision, err := c.lister.List(selector.AsSelector())
+	if err != nil {
+		logrus.WithError(err).Errorf("Cannot list prowjob with selector %v", selector)
+		return err
 	}
 
-	message := fmt.Sprintf("Job %s finished with %s\n Gubernator URL: %s", pj.Spec.Job, pj.Status.State, pj.Status.URL)
+	// generate an aggregated report:
+	total := len(pjsOnRevision)
+	success := 0
+	message := ""
+
+	for _, pjOnRevision := range pjsOnRevision {
+		if pjOnRevision.Status.PrevReportStates[c.GetName()] == pjOnRevision.Status.State {
```

Comment:
Why are we setting the value here to `.Status.State` (success, failure, pending) rather than a bool? 

It also seems like it won't work correctly, for example in a situation where for some reason a third test starts after the first two pass... we'd want to leave a comment again after all three pass. but we won't here unless the third one has a different result than the first two.

Reply:
If we ever want to support `/retest` from a comment, we probably need to somehow invalidate the previous version of the job. But for now all tests for the same patchset will be triggered together.

Reply:
I don't see how this is related.

Reply:
otherwise I don't see a situation when `for some reason a third test starts after the first two pass`

Reply:
If we are worried about locks, we should also be worried about edge cases where jobs 1 and 2 fail before we manage to schedule job 3, right?

Reply:
so ideally, the adapter will create all prowjobs on the same revision... in the case of something like kube-apiserver delayed for 5min this case will happen... I would vote we don't need to worry about the case here.. (delay schedule the pod is fine as long as prowjobs are created)